benchexec/container.py

```diff
@@ -531,29 +542,50 @@ def duplicate_mount_hierarchy(mount_base, temp_base, work_base, dir_modes):
             work_path = work_base + b"/" + str(overlay_count).encode()
             os.makedirs(temp_path, exist_ok=True)
             os.makedirs(work_path, exist_ok=True)
-            try:
-                # Previous mount in this place not needed if replaced with overlay dir.
-                libc.umount(mount_path)
-            except OSError as e:
-                logging.debug(e)
-            try:
-                make_overlay_mount(mount_path, mountpoint, temp_path, work_path)
-            except OSError as e:
-                mp = mountpoint.decode()
-                raise OSError(
-                    e.errno,
-                    f"Creating overlay mount for '{mp}' failed: {os.strerror(e.errno)}. "
-                    f"Please use other directory modes, "
-                    f"for example '--read-only-dir {shlex.quote(mp)}'.",
-                )
+            if os.path.ismount(mount_path):
+                try:
+                    # Previous mount in this place not needed if replaced with overlay dir.
+                    libc.umount(mount_path)
+                except OSError as e:
+                    logging.debug(e)
+
+            if use_fuse and fuse_overlay_mount_path:
+                fuse_mount_path = fuse_overlay_mount_path + mountpoint
+                make_bind_mount(fuse_mount_path, mount_path)
+            else:
+                try:
+                    make_overlay_mount(mount_path, mountpoint, temp_path, work_path)
+                except OSError as e:
+                    # Resort to fuse-overlayfs if kernel overlayfs is not available.
+                    # This part of the code (using fuse-overlayfs as a fallback) is intentionally
+                    # kept as a workaround for triple-nested execution with kernel overlayfs.
+                    mp = mountpoint.decode()
+                    if not fuse_overlay_mount_path:
+                        fuse_overlay_mount_path = setup_fuse_overlay(
+                            temp_base, work_base
+                        )
+                    if not fuse_overlay_mount_path:
+                        logging.warning("fuse-overlayfs is not available.")
+                        raise OSError(
+                            e.errno,
+                            f"Failed to create overlay mount for '{mp}': {os.strerror(e.errno)}. "
+                            f"Consider using alternative directory modes, such as '--read-only-dir {shlex.quote(mp)}'.",
+                        ) from e
+
+                    logging.debug(
+                        f"Fallback to fuse-overlayfs for overlay mount at '{mp}'."
```

Comment:
Nitpicking: While we prefer f-strings in general, for logging they have the disadvantage that the string creation is performed eagerly even if we do not need it. But `logging` has builtin support for `%s` formatting, which is lazy. So in general we prefer that for log messages (`logging.debug("...%s...", mp)`).

Reply:
I've learned something new! Thanks